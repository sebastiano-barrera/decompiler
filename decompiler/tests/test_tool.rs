//! Tests based on "test_tool"
//!
//! These are pretty weak, but they work on real code. They only check that the
//! decompiler doesn't panic, and they write the decompiler's output as text to
//! a version-controlled file, just so that it's somewhere easy to diff and hard
//! to forget updating.

use include_dir::{include_dir, Dir};
use insta::assert_snapshot;

use std::{
    path::Path,
    sync::{Mutex, MutexGuard, OnceLock},
};

static DATA_DIR: Dir<'_> = include_dir!("$CARGO_MANIFEST_DIR/test-data/");

macro_rules! case {
    ($case:ident, $exe:expr, $func_name:ident) => {
        #[test_log::test]
        #[allow(non_snake_case)]
        fn $case() {
            let function_name = stringify!($func_name);
            let out = $exe.process_function(function_name);
            crate::assert_snapshot!(out);
        }
    };
}

macro_rules! tests_in_binary {
    ($group:ident, $exe_path:expr; $($funcs:ident),* $(,)?) => {
        mod $group {
            use crate::Exe;
            static EXE: Exe = Exe::new($exe_path);

            $(case!($funcs, EXE, $funcs);)*
        }
    };
}

struct Exe {
    once_lock: OnceLock<Mutex<decompiler::Executable<'static>>>,
    path: &'static str,
}
impl Exe {
    const fn new(path: &'static str) -> Self {
        Exe {
            once_lock: OnceLock::new(),
            path,
        }
    }

    fn get_or_init(&self) -> MutexGuard<'_, decompiler::Executable<'static>> {
        let mutex = self.once_lock.get_or_init(|| {
            let rel_path = Path::new(self.path);
            let raw = DATA_DIR.get_file(rel_path).unwrap().contents();
            let exe = decompiler::Executable::parse(raw).unwrap();
            Mutex::new(exe)
        });
        mutex.lock().unwrap()
    }

    fn process_function(&self, function_name: &str) -> String {
        use std::fmt::Write;

        let mut exe = self.get_or_init();

        let df = exe
            .decompile_function(function_name)
            .expect("decompiling function");

        let mut log_buf = String::new();

        if let Some(mil) = df.mil() {
            writeln!(log_buf, " --- mil").unwrap();
            writeln!(log_buf, "{:?}\n", mil).unwrap();
        }

        if let Some(ssa) = df.ssa_pre_xform() {
            writeln!(log_buf, " --- ssa pre-xform").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();
        }

        if let Some(ssa) = df.ssa() {
            writeln!(log_buf, " --- cfg").unwrap();
            let cfg = ssa.cfg();
            writeln!(log_buf, "  entry: {:?}", cfg.direct().entry_bid()).unwrap();
            for bid in cfg.block_ids() {
                let regs: Vec<_> = ssa.block_regs(bid).collect();
                writeln!(
                    log_buf,
                    "  {:?} -> {:?} {:?}",
                    bid,
                    cfg.block_cont(bid),
                    regs
                )
                .unwrap();
            }
            write!(log_buf, "  domtree:\n    ").unwrap();

            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            cfg.dom_tree().dump(pp).unwrap();
            writeln!(log_buf).unwrap();

            writeln!(log_buf, " --- ssa").unwrap();
            writeln!(log_buf, "{:?}\n", ssa).unwrap();

            writeln!(log_buf, " --- ast").unwrap();
            let mut ast = decompiler::Ast::new(&ssa, exe.types());
            let mut pp_buf = decompiler::pp::FmtAsIoUTF8(&mut log_buf);
            let pp = &mut decompiler::pp::PrettyPrinter::start(&mut pp_buf);
            ast.pretty_print(pp).unwrap();
        }

        log_buf
    }
}

tests_in_binary!(
    exe_composite_type, "ty/test_composite_type.so";
    list_len
);
tests_in_binary!(exe_capstone, "integration/sample_capstone"; main);
tests_in_binary!(exe_matrix, "integration/sample_matrix"; sum_matrix);
tests_in_binary!(
    redis_server, "integration/redis-server";
    ctl_lookup,
    listNext,
    listTypePush,
    geoArrayCreate,
    // regression: infinite loop
    processAnnotations,
    redisReconnect,
    redisConnectWithOptions,
);
