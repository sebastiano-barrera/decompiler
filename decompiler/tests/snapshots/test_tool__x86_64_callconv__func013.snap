---
source: decompiler/tests/test_tool.rs
expression: out
---
 --- mil
program  59 instrs
    0 r12: UnknownUnsized <- Ancestral(AncestralName("stack_bottom"))
    1 r2: UnknownUnsized <- Ancestral(AncestralName("CF"))
    2 r3: UnknownUnsized <- Ancestral(AncestralName("PF"))
    3 r4: UnknownUnsized <- Ancestral(AncestralName("AF"))
    4 r5: UnknownUnsized <- Ancestral(AncestralName("ZF"))
    5 r6: UnknownUnsized <- Ancestral(AncestralName("SF"))
    6 r7: UnknownUnsized <- Ancestral(AncestralName("TF"))
    7 r8: UnknownUnsized <- Ancestral(AncestralName("IF"))
    8 r9: UnknownUnsized <- Ancestral(AncestralName("DF"))
    9 r10: UnknownUnsized <- Ancestral(AncestralName("OF"))
   10 r11: UnknownUnsized <- Ancestral(AncestralName("RBP"))
   11 r12: UnknownUnsized <- Ancestral(AncestralName("RSP"))
   12 r13: UnknownUnsized <- Ancestral(AncestralName("RIP"))
   13 r14: UnknownUnsized <- Ancestral(AncestralName("RDI"))
   14 r15: UnknownUnsized <- Ancestral(AncestralName("RSI"))
   15 r16: UnknownUnsized <- Ancestral(AncestralName("RAX"))
   16 r17: UnknownUnsized <- Ancestral(AncestralName("RBX"))
   17 r18: UnknownUnsized <- Ancestral(AncestralName("RCX"))
   18 r19: UnknownUnsized <- Ancestral(AncestralName("RDX"))
   19 r20: UnknownUnsized <- Ancestral(AncestralName("R8"))
   20 r21: UnknownUnsized <- Ancestral(AncestralName("R9"))
   21 r22: UnknownUnsized <- Ancestral(AncestralName("R10"))
   22 r23: UnknownUnsized <- Ancestral(AncestralName("R11"))
   23 r24: UnknownUnsized <- Ancestral(AncestralName("R12"))
   24 r25: UnknownUnsized <- Ancestral(AncestralName("R13"))
   25 r26: UnknownUnsized <- Ancestral(AncestralName("R14"))
   26 r27: UnknownUnsized <- Ancestral(AncestralName("R15"))
   27 r28: UnknownUnsized <- Ancestral(AncestralName("ZMM0"))
   28 r29: UnknownUnsized <- Ancestral(AncestralName("ZMM1"))
   29 r30: UnknownUnsized <- Ancestral(AncestralName("ZMM2"))
   30 r31: UnknownUnsized <- Ancestral(AncestralName("ZMM3"))
   31 r32: UnknownUnsized <- Ancestral(AncestralName("ZMM4"))
   32 r33: UnknownUnsized <- Ancestral(AncestralName("ZMM5"))
   33 r34: UnknownUnsized <- Ancestral(AncestralName("ZMM6"))
   34 r35: UnknownUnsized <- Ancestral(AncestralName("ZMM7"))
   35 r36: UnknownUnsized <- Ancestral(AncestralName("ZMM8"))
   36 r37: UnknownUnsized <- Ancestral(AncestralName("ZMM9"))
   37 r38: UnknownUnsized <- Ancestral(AncestralName("ZMM10"))
   38 r39: UnknownUnsized <- Ancestral(AncestralName("ZMM11"))
   39 r40: UnknownUnsized <- Ancestral(AncestralName("ZMM12"))
   40 r41: UnknownUnsized <- Ancestral(AncestralName("ZMM13"))
   41 r42: UnknownUnsized <- Ancestral(AncestralName("ZMM14"))
   42 r43: UnknownUnsized <- Ancestral(AncestralName("ZMM15"))
   43 r45: UnknownUnsized <- Ancestral(AncestralName("arg0"))
   44 r46: UnknownUnsized <- Part { src: r45, offset: 0, size: 8 }
   45 r47: UnknownUnsized <- Part { src: r28, offset: 8, size: 56 }
   46 r28: UnknownUnsized <- Concat { lo: r46, hi: r47 }
   47 r48: UnknownUnsized <- Part { src: r45, offset: 8, size: 8 }
   48 r49: UnknownUnsized <- Part { src: r29, offset: 8, size: 56 }
   49 r29: UnknownUnsized <- Concat { lo: r48, hi: r49 }
0x401138:
   50 r46: UnknownUnsized <- Part { src: r29, offset: 0, size: 16 }
   51 r47: UnknownUnsized <- Part { src: r28, offset: 16, size: 48 }
   52 r28: UnknownUnsized <- Concat { lo: r46, hi: r47 }
0x40113b:
   53 r45: UnknownUnsized <- Void
   54 r46: UnknownUnsized <- Part { src: r28, offset: 0, size: 8 }
   55 r45: UnknownUnsized <- Concat { lo: r45, hi: r46 }
   56 r45: UnknownUnsized <- Part { src: r45, offset: 0, size: 8 }
   57 r47: UnknownUnsized <- SetReturnValue(r45)
   58 r47: UnknownUnsized <- Control(Ret)


 --- ssa pre-xform
ssa program  59 instrs
.B0:    ;;  → Always(Return).
         r27: unknown <UnknownUnsized> <- Ancestral(AncestralName("ZMM0"))
         r28: unknown <UnknownUnsized> <- Ancestral(AncestralName("ZMM1"))
  (  2)  r43: unknown <UnknownUnsized> <- Ancestral(AncestralName("arg0"))
         r44: unknown <UnknownUnsized> <- Part { src: r43, offset: 0, size: 8 }
         r45: unknown <UnknownUnsized> <- Part { src: r27, offset: 8, size: 56 }
         r46: unknown <UnknownUnsized> <- Concat { lo: r44, hi: r45 }
         r47: unknown <UnknownUnsized> <- Part { src: r43, offset: 8, size: 8 }
         r48: unknown <UnknownUnsized> <- Part { src: r28, offset: 8, size: 56 }
         r49: unknown <UnknownUnsized> <- Concat { lo: r47, hi: r48 }
         r50: unknown <UnknownUnsized> <- Part { src: r49, offset: 0, size: 16 }
         r51: unknown <UnknownUnsized> <- Part { src: r46, offset: 16, size: 48 }
         r52: unknown <UnknownUnsized> <- Concat { lo: r50, hi: r51 }
         r54: unknown <UnknownUnsized> <- Part { src: r52, offset: 0, size: 8 }
         r55: unknown <UnknownUnsized> <- Concat { lo: r53, hi: r54 }
         r56: unknown <UnknownUnsized> <- Part { src: r55, offset: 0, size: 8 }
         r57: unknown <UnknownUnsized> <- SetReturnValue(r56)


 --- cfg
  entry: BlockID(0)
  BlockID(0) -> Always(Return) [r43, r56, r57]
  domtree:
    BlockID(0)

 --- ssa
ssa program  59 instrs
.B0:    ;;  → Always(Return).
         r43: unknown <UnknownUnsized> <- Ancestral(AncestralName("arg0"))
         r56: unknown <UnknownUnsized> <- Part { src: r43, offset: 8, size: 8 }
         r57: unknown <UnknownUnsized> <- SetReturnValue(r56)


 --- ast

T0: {
  ;
  return pre:arg0[16 .. 8]
}
